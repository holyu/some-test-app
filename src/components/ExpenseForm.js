import React from "react";
import DatePicker from "react-date-picker/dist/entry.nostyle";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense
        ? (props.expense.amount / 100).toString().replace(".", ",")
        : "",
      createdAt: props.expense ? new Date(props.expense.createdAt) : new Date(),
      error: undefined
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = e => {
    let amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\,\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount."
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount.replace(",", "."), 10) * 100,
        createdAt: this.state.createdAt.getTime(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <DatePicker
          className="datepicker no-clear"
          onChange={this.onDateChange}
          value={this.state.createdAt}
          returnValue="start"
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          {this.props.expense ? (
            <button className="button">Save Expense</button>
          ) : (
            <button className="button">Add Expense</button>
          )}
        </div>
      </form>
    );
  }
}
