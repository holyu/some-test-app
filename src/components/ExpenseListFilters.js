import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import DatePicker from "react-date-picker/dist/entry.nostyle";

export class ExpenseListFilters extends React.Component {
  onStartDatesChange = startDate => {
    this.props.setStartDate(startDate);
  };
  onEndDatesChange = endDate => {
    this.props.setEndDate(endDate);
  };
  onTextChange = e => this.props.setTextFilter(e.target.value);
  onSortChange = e => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else if (e.target.value === "date") {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <DatePicker
              className="datepicker"
              onChange={this.onStartDatesChange}
              value={this.props.filters.startDate}
              returnValue="start"
            />
          </div>
          <div className="input-group__item">
            <DatePicker
              className="datepicker"
              onChange={this.onEndDatesChange}
              value={this.props.filters.endDate}
              returnValue="end"
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
