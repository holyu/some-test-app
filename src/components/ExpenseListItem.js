import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">
        {new Intl.DateTimeFormat("pl-PL", {
          year: "numeric",
          month: "short",
          day: "numeric"
        }).format(createdAt)}
      </span>
    </div>
    <h3 className="list-item__data">
      {new Intl.NumberFormat("pl-PL", {
        style: "currency",
        currency: "PLN"
      }).format(amount / 100)}
    </h3>
  </Link>
);

export default ExpenseListItem;
