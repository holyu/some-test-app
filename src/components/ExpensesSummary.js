import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = ({ expensesTotal, expensesCount }) => {
  const expensedText = expensesCount === 1 ? "expense" : "expenses";
  const formattedTotal = new Intl.NumberFormat("pl-PL", {
    style: "currency",
    currency: "PLN"
  }).format(expensesTotal / 100);
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expensesCount}</span> {expensedText} totalling{" "}
          <span>{formattedTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const allVisibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: selectExpensesTotal(allVisibleExpenses),
    expensesCount: allVisibleExpenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
