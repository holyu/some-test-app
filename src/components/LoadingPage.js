import React from "react";
import loaderImg from "../images/loader.gif";

export const LoadingPage = () => (
  <div className="loader">
    <img className="loader__image" src={loaderImg} />
  </div>
);

export default LoadingPage;
