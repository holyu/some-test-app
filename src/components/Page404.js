import React from "react";
import { Link } from "react-router-dom";

export default () => (
  <div>
    This is 404! - <Link to="/">Home</Link>
  </div>
);
