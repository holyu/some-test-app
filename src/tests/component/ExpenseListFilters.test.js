import React from "react";
import { shallow } from "enzyme";
import { ExpenseListFilters } from "../../components/ExpenseListFilters";
import { filters, altFilters } from "../fixtures/filters";

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, wrapper;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByDate = jest.fn();
  sortByAmount = jest.fn();
  setStartDate = jest.fn();
  setEndDate = jest.fn();
  wrapper = shallow(
    <ExpenseListFilters
      filters={filters}
      setTextFilter={setTextFilter}
      sortByDate={sortByDate}
      sortByAmount={sortByAmount}
      setStartDate={setStartDate}
      setEndDate={setEndDate}
    />
  );
});

test("should render ExpenseListFilters correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseListFilters with alt date correctly", () => {
  wrapper.setProps({ filters: altFilters });
  expect(wrapper).toMatchSnapshot();
});

test("should handle text change", () => {
  wrapper
    .find("input")
    .at(0)
    .simulate("change", { target: { value: altFilters.text } });
  expect(setTextFilter).toHaveBeenLastCalledWith(altFilters.text);
});

test("should sort by date", () => {
  wrapper.setProps({ filters: altFilters });
  wrapper
    .find("select")
    .simulate("change", { target: { value: filters.sortBy } });
  expect(sortByDate).toHaveBeenCalled();
});

test("should sort by amount", () => {
  wrapper
    .find("select")
    .simulate("change", { target: { value: altFilters.sortBy } });
  expect(sortByAmount).toHaveBeenCalled();
});

test("should handle start date changes", () => {
  wrapper.setProps({ filters: altFilters });
  const startDate = altFilters.startDate;
  wrapper
    .find("DatePicker")
    .at(0)
    .prop("onChange")(startDate);
  expect(setStartDate).toHaveBeenLastCalledWith(startDate);
});

test("should handle end date changes", () => {
  wrapper.setProps({ filters: altFilters });
  const endDate = altFilters.endDate;
  wrapper
    .find("DatePicker")
    .at(1)
    .prop("onChange")(endDate);
  expect(setEndDate).toHaveBeenLastCalledWith(endDate);
});
