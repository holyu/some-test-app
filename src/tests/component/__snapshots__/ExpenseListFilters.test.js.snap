// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search expenses"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item"
    >
      <DatePicker
        calendarIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </g>
          </svg>
        }
        className="datepicker"
        clearIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </g>
          </svg>
        }
        isOpen={null}
        onChange={[Function]}
        returnValue="start"
        value={null}
      />
    </div>
    <div
      className="input-group__item"
    >
      <DatePicker
        calendarIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </g>
          </svg>
        }
        className="datepicker"
        clearIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </g>
          </svg>
        }
        isOpen={null}
        onChange={[Function]}
        returnValue="end"
        value={null}
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="date"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseListFilters with alt date correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search expenses"
        type="text"
        value="bills"
      />
    </div>
    <div
      className="input-group__item"
    >
      <DatePicker
        calendarIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </g>
          </svg>
        }
        className="datepicker"
        clearIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </g>
          </svg>
        }
        isOpen={null}
        onChange={[Function]}
        returnValue="start"
        value={1970-01-01T00:00:00.000Z}
      />
    </div>
    <div
      className="input-group__item"
    >
      <DatePicker
        calendarIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </g>
          </svg>
        }
        className="datepicker"
        clearIcon={
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              stroke="black"
              strokeWidth="2"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </g>
          </svg>
        }
        isOpen={null}
        onChange={[Function]}
        returnValue="end"
        value={1970-01-01T00:01:10.000Z}
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="amount"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
  </div>
</div>
`;
